Iterators and Access functions for edges

Rohit-Kakodkar opened the pull request:
## Description

- [x] Adds chunk edge iterator
   - [x] Adds tests for chunk edge iterator
- [x] Adds edge intersection iterator
   - [x] Adds tests for edge intersection iterator
- [x] Adds access functions for chunk field

## Issue Number

If there is an issue created for these changes, link it here

## Checklist

Please make sure to check developer documentation on specfem docs.

- [ ] I ran the code through pre-commit to check style
- [ ] **THE DOCUMENTATION BUILDS WITHOUT WARNINGS/ERRORS**
- [ ] I have added labels to the PR (see right hand side of the PR page)
- [ ] My code passes all the integration tests
- [ ] I have added sufficient unittests to test my changes
- [ ] I have added/updated documentation for the changes I am proposing
- [ ] I have updated CMakeLists to ensure my code builds
- [ ] My code builds across all platforms

Rohit-Kakodkar committed:
Iterators and Access functions for edges

- [x] Adds chunk edge iterator
   - [x] Adds tests for chunk edge iterator
- [x] Adds edge intersection iterator
   - [x] Adds tests for edge intersection iterator
- [x] Adds access functions for chunk field

Rohit-Kakodkar committed:
Merge branch 'devel-dg' into issue-1150

Rohit-Kakodkar committed:
Fixed boost library linker

int-ptr-ptr reviewed (approved):
While I think the binary was-everything-hit-exactly-once test is fine for now, I think in the future, we may want to also check hit order -- though, that may be within the scope of tests post-assembly (do `x` and `z` match between self and coupled edge, etc.).

lsawade reviewed (approved):
LGTM, comments are at your own discretion

lsawade commented:
#1160 may fix unit-test

Rohit-Kakodkar committed:
Refactor `FOR_EACH` macro to incorporate interface tags

- [x] Added `ENUM_ID` to enable strong type checking when checking tags within `FOR_EACH` macro. The macro `_CHECK_ENUM` performs the strong type checking when comparting 2 enums.
- [x] Added edge definition macros to enable all types of edges

Rohit-Kakodkar committed:
Updated edge definitions and added constexpr tags

Rohit-Kakodkar committed:
Added Edges to tag 4 macro

Rohit-Kakodkar committed:
Merge branch 'issue-1165' into issue-1167

Rohit-Kakodkar committed:
Create edge types to sort all edges with tags

- [x] Updates mesh containers to hold connecting edges
- [x] Adds weakly conforming connection to represent coupled interfaces
- [x] Updates mesh reader
- [x] Adds edge types container to sort edges for kernel computation

Rohit-Kakodkar committed:
Merge branch 'issue-1150' into issue-1165

Rohit-Kakodkar committed:
Merge branch 'issue-1165' into issue-1167

Rohit-Kakodkar committed:
Added comments to deprecate adjancency graph

Rohit-Kakodkar committed:
Merge pull request #1179 from PrincetonUniversity/issue-1167

Create edge types to sort all edges with tags

Rohit-Kakodkar committed:
Merge pull request #1166 from PrincetonUniversity/issue-1165

Refactor `FOR_EACH` macro to incorporate interface tags

Rohit-Kakodkar committed:
Reimplement coupled interfaces

- [x] Updates coupled interfaces struct
- [x] Adds several new point types (point edge index, point interface index, point coupled interface)
- [x] Adds coupled interfaces 2 container which will replace current coupled interface container

Rohit-Kakodkar committed:
Minor fixes to mesh interface container

Rohit-Kakodkar committed:
Implements new kokkos kernel to compute coupling

- [x] New compute coupling kokkos kernel using ChunkEdgePolicy
- [x] Media specific compute coupling (elastic-acoustic & acoustic-elastic)
- [x] Adds unit-tests

Rohit-Kakodkar committed:
Merge pull request #1189 from PrincetonUniversity/issue-1167

Reimplement coupled interfaces using macros

Rohit-Kakodkar committed:
Added missing headers

Rohit-Kakodkar committed:
Updated calls to coupling kernels

Rohit-Kakodkar committed:
Merge pull request #1194 from PrincetonUniversity/issue-1167

Implements new kokkos kernel to compute coupling

Rohit-Kakodkar committed:
Fixed bugs in compute coupling

Rohit-Kakodkar committed:
Deprecated old coupled interfaces implementation

- [x] Remove old coupled interfaces implementation
- [x] Applied dirichlet boundary conditions

Rohit-Kakodkar committed:
Renames coupled_interfaces2 -> coupled_interfaces

Rohit-Kakodkar committed:
Added documentation

Rohit-Kakodkar committed:
Commented display section

Rohit-Kakodkar committed:
Merge pull request #1201 from PrincetonUniversity/issue-1172

Update compute coupling calls

Rohit-Kakodkar committed:
Merge branch 'devel' into issue-1150

Rohit-Kakodkar committed:
minor fix

Rohit-Kakodkar committed:
Fixed some access functions

Rohit-Kakodkar committed:
Minor bug fix for NVIDIA

Rohit-Kakodkar committed:
Update tests/unit-tests/policies/chunked_edge.cpp

Co-authored-by: Lucas Sawade <lsawade@princeton.edu>

Rohit-Kakodkar committed:
Update tests/unit-tests/policies/chunked_edge.cpp

Co-authored-by: Lucas Sawade <lsawade@princeton.edu>

Rohit-Kakodkar committed:
Revert "Minor bug fix for NVIDIA"

This reverts commit 55d2717ec950e6235e6ecf4d83a8a5e220319494.

Rohit-Kakodkar committed:
Fixed NVIDIA tests
